### ISSUE TITLE
[HOLD for payment 2023-03-06] [HOLD for payment 2023-02-10] [$1000]  Timezone changes to some 1st option in the list while pressing the enter key

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#14314]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Go to [new.expensify.com](http://new.expensify.com/)
2. Login with any account
3. Go to Profile and click on timezone
4. Change timezone to some other regions and go back
5. Again click on timezone and now hit the enter key
6. Notice that timezone changes to some other region on pressing the enter key (This doesn’t happen on android/ios while pressing enter on the keyboard. It just happens on web)

## Expected Result:
When showing the timezone list the timezone that was set should be selected and visible and pressing the enter key should select the current selected timezone (in case no interaction from the user was made then nothing should change)

## Actual Result:
When opening the timezone selection the actual set timezone does not apperar selected, and when pressing enter the first item on the list is seleced instead.

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] MacOS / Chrome / Safari

**Version Number:** 1.2.54-2
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/212493695-0cfafa96-585f-47fe-8f9a-44172b6fb278.mp4

https://user-images.githubusercontent.com/43996225/212493698-d1b60f7f-6080-41e3-a5fa-b8cdf6d6bcc8.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @priya-zha
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1673705115960529

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01faba6d65b765cfed</li>
        <li>Upwork Job ID: 1615516451621584896</li>
        <li>Last Price Increase: 2023-01-18</li>
    </ul>
</details>

**Comment by kadiealexander** [comment]
### Bug0 Triage Checklist
_Note: see [this SO](https://stackoverflow.com/c/expensify/questions/14418) for more information._
- [x] The "bug" is actually a bug 
- [x] The bug is not a duplicate report of an existing GH.
  - If it is, close the GH and add any novel details to the original GH instead 
- [x] The bug is reproducible, following the reproduction steps.
  - [x] If the GH doesn’t have steps to reliably reproduce the bug and you figure it out, then please add them.
  - [x] If you’re unable to reproduce the bug, add the `Needs reproduction` label. 
  - [x] Comment on the issue outlining the steps you took to try to reproduce the bug, your results and tag the issue reporter and the Applause QA member who created the issue. Ask them to clarify reproduction steps and/or to check the reproduction steps again. Close issue. 
- [x] The GH template is filled out as fully as possible
  - The GH body and title are clear (ie. could an external contributor understand it and work on it?) 
- [x] The GH was created by an Expensify employee or a QA tester 
- [x] If the bug is an OldDot issue, you should decide whether it is widespread enough to justify fixing or it is better to wait for reunification. If it’s better to wait, close the GH & provide this justification 
- [x] If there's a link to Slack, check the discussion to see if we decided not to fix it 
- [x] Decide if the GH should be resolved by an External contributor or Internal engineer, add the appropriate label

Reproducible, timezone selects the first option in the list when hitting enter, even when a different timezone is written in the text box:

https://user-images.githubusercontent.com/59587260/213055796-a8ead17e-ceb6-48e5-9454-9f828db5e36e.mp4

**Comment by tienifr** [comment]
# Proposal

This happens due to we're focusing on the first element always upon rendering the OptionList, and when we press enter it will select the currently focused element. To fix this we need to focus on the currently selected element, if an element was selected prior to rendering the OptionList

```diff
diff --git a/src/components/OptionsSelector/BaseOptionsSelector.js b/src/components/OptionsSelector/BaseOptionsSelector.js
index c473c426d8..b1d5e01957 100755
--- a/src/components/OptionsSelector/BaseOptionsSelector.js
+++ b/src/components/OptionsSelector/BaseOptionsSelector.js
@@ -39,9 +39,17 @@ class BaseOptionsSelector extends Component {
         this.relatedTarget = null;
 
         const allOptions = this.flattenSections();
+
+        const selectedOptionIndex = _.findIndex(allOptions, option => option.text === this.props.value);
+
+        let focusedIndex = this.props.shouldTextInputAppearBelowOptions ? allOptions.length : 0;
+        if (selectedOptionIndex >= 0) {
+            focusedIndex = selectedOptionIndex;
+        }
+
         this.state = {
             allOptions,
-            focusedIndex: this.props.shouldTextInputAppearBelowOptions ? allOptions.length : 0,
+            focusedIndex,
         };
     }
 
diff --git a/src/pages/settings/Profile/TimezoneSelectPage.js b/src/pages/settings/Profile/TimezoneSelectPage.js
index 7a89ac8ce7..7fc47ee5e8 100644
--- a/src/pages/settings/Profile/TimezoneSelectPage.js
+++ b/src/pages/settings/Profile/TimezoneSelectPage.js
@@ -61,6 +61,9 @@ class TimezoneSelectPage extends Component {
      */
     saveSelectedTimezone({text}) {
         PersonalDetails.updateSelectedTimezone(text);
+        this.setState({
+            timezoneInputText: text,
+        });
     }
 
     /**
```

We might need to fix this in other places as well with the same update

**Comment by tienifr** [comment]
If we don't want it to affect anything else yet, then can introduce a new params in the `BaseOptionsSelector` to check for the focused index rather than using the `props.value`

**Comment by tienifr** [comment]
Working well after both proposals:

https://user-images.githubusercontent.com/113963320/213066187-44810b46-22e5-42ea-a20b-58a6b5ba6cfa.mov

**Comment by bernhardoj** [comment]
### **Proposal**

If we set the `focusedIndex` based on previous selection, when we press arrow up/down, it will jump to the previous selected item. I think what we need to do here is to show that the first item is the focused item, like the other list. At timezone selection page, we forget to pass the indexOffset to the sections making the option item focus style didn't show.

```diff
diff --git a/src/pages/settings/Profile/TimezoneSelectPage.js b/src/pages/settings/Profile/TimezoneSelectPage.js
index 7a89ac8ce7..f3d4a229c3 100644
--- a/src/pages/settings/Profile/TimezoneSelectPage.js
+++ b/src/pages/settings/Profile/TimezoneSelectPage.js
@@ -88,7 +88,7 @@ class TimezoneSelectPage extends Component {
                     onChangeText={this.filterShownTimezones}
                     onSelectRow={this.saveSelectedTimezone}
                     optionHoveredStyle={styles.hoveredComponentBG}
-                    sections={[{data: this.state.timezoneOptions}]}
+                    sections={[{data: this.state.timezoneOptions, indexOffset: 0}]}
                     shouldHaveOptionSeparator
                 />
             </ScreenWrapper>

```

**Result**

https://user-images.githubusercontent.com/50919443/213070775-c4830e0f-9ab9-4b45-8448-7868e316327e.mov

**Comment by tienifr** [comment]
> when we press arrow up/down, it will jump to the previous selected item

If we want to avoid this minor issue, then we can default scroll to the selected element when we open the list (in `componentDidMount`). `animated: false` can also be added if we want it to snap into the position rather than scrolling into it.

```diff
diff --git a/src/components/OptionsSelector/BaseOptionsSelector.js b/src/components/OptionsSelector/BaseOptionsSelector.js
index c473c426d8..60b53f294c 100755
--- a/src/components/OptionsSelector/BaseOptionsSelector.js
+++ b/src/components/OptionsSelector/BaseOptionsSelector.js
@@ -93,6 +103,8 @@ class BaseOptionsSelector extends Component {
         } else {
             this.textInput.focus();
         }
+
+        this.scrollToIndex(this.state.focusedIndex);
     }

```

I've tested the timezone selection on Slack and these are consistent with Slack's behavior:
- When on the timezone list and press enter, the selected timezone should stay the same (rather than selecting the 1st timezone)
- When opening the timezone list for the first time, the selected timezone will show on the screen. We do not need to scroll to view the selected timezone.

<img width="567" alt="Screenshot 2023-01-18 at 10 46 39" src="https://user-images.githubusercontent.com/113963320/213078784-68e4cf72-8d16-4e7c-95f9-13027c0cfcad.png">

Think of the user's viewpoint, if he changes the timezone it's likely will be surrounding timezone in `Asia` for example, he might not want to scroll all the way just to see the `Asia` timezone, it should appear by default.

**Comment by tienifr** [comment]
There is another place that is also having this problem which may also need to be fixed

https://user-images.githubusercontent.com/113963320/213079379-c77ad986-9b7f-4c04-a7fb-254cbbbababb.mov

**Comment by s77rt** [comment]
I think pressing enter should select the 1st option, that's the expected behaviour not a bug. The bug is that from the user perspective this is confusing because we are missing the focus state indication on the current focused row. I think all what we need to do here is apply @bernhardoj proposal [here](https://github.com/Expensify/App/issues/14314#issuecomment-1386404745)

**Comment by s77rt** [comment]
### Proposal
```diff
diff --git a/src/components/OptionsSelector/BaseOptionsSelector.js b/src/components/OptionsSelector/BaseOptionsSelector.js
index c473c426d8..904de0450d 100755
--- a/src/components/OptionsSelector/BaseOptionsSelector.js
+++ b/src/components/OptionsSelector/BaseOptionsSelector.js
@@ -41,7 +41,7 @@ class BaseOptionsSelector extends Component {
         const allOptions = this.flattenSections();
         this.state = {
             allOptions,
-            focusedIndex: this.props.shouldTextInputAppearBelowOptions ? allOptions.length : 0,
+            focusedIndex: props.shouldTextInputAppearBelowOptions ? allOptions.length : 0,
         };
     }
 
@@ -84,6 +84,10 @@ class BaseOptionsSelector extends Component {
             true,
         );
 
+        if (this.props.focusedItemKey) {
+            this.updateFocusedIndex(_.findIndex(this.state.allOptions, option => option.keyForList === this.props.focusedItemKey), false);
+        }
+
         if (!this.props.autoFocus) {
             return;
         }
@@ -166,17 +170,19 @@ class BaseOptionsSelector extends Component {
 
     /**
      * @param {Number} index
+     * @param {Boolean} animated
      */
-    updateFocusedIndex(index) {
-        this.setState({focusedIndex: index}, () => this.scrollToIndex(index));
+    updateFocusedIndex(index, animated = true) {
+        this.setState({focusedIndex: index}, () => this.scrollToIndex(index, animated));
     }
 
     /**
      * Scrolls to the focused index within the SectionList
      *
      * @param {Number} index
+     * @param {Boolean} animated
      */
-    scrollToIndex(index) {
+    scrollToIndex(index, animated = true) {
         const option = this.state.allOptions[index];
         if (!this.list || !option) {
             return;
@@ -195,7 +201,7 @@ class BaseOptionsSelector extends Component {
             }
         }
 
-        this.list.scrollToLocation({sectionIndex: adjustedSectionIndex, itemIndex});
+        this.list.scrollToLocation({sectionIndex: adjustedSectionIndex, itemIndex, animated});
     }
 
     /**
diff --git a/src/components/OptionsSelector/optionsSelectorPropTypes.js b/src/components/OptionsSelector/optionsSelectorPropTypes.js
index 306b89a26e..51691fb41e 100644
--- a/src/components/OptionsSelector/optionsSelectorPropTypes.js
+++ b/src/components/OptionsSelector/optionsSelectorPropTypes.js
@@ -89,6 +89,9 @@ const propTypes = {
 
     /** Whether to show a line separating options in list */
     shouldHaveOptionSeparator: PropTypes.bool,
+
+    /** Initially focused item key */
+    focusedItemKey: PropTypes.string,
 };
 
 const defaultProps = {
@@ -113,6 +116,7 @@ const defaultProps = {
     disableArrowKeysActions: false,
     isDisabled: false,
     shouldHaveOptionSeparator: false,
+    focusedItemKey: '',
 };
 
 export {propTypes, defaultProps};
diff --git a/src/pages/settings/Profile/TimezoneSelectPage.js b/src/pages/settings/Profile/TimezoneSelectPage.js
index 7a89ac8ce7..2018ca275a 100644
--- a/src/pages/settings/Profile/TimezoneSelectPage.js
+++ b/src/pages/settings/Profile/TimezoneSelectPage.js
@@ -88,7 +88,8 @@ class TimezoneSelectPage extends Component {
                     onChangeText={this.filterShownTimezones}
                     onSelectRow={this.saveSelectedTimezone}
                     optionHoveredStyle={styles.hoveredComponentBG}
-                    sections={[{data: this.state.timezoneOptions}]}
+                    sections={[{data: this.state.timezoneOptions, indexOffset: 0}]}
+                    focusedItemKey={this.currentSelectedTimezone}
                     shouldHaveOptionSeparator
                 />
             </ScreenWrapper>
```

### RCA
This seems more like a feature request: scroll to selected item
1. I have added a prop `focusedItemKey` to initially focus an item based on matching `keyForList`
2. Fixed the missing focus (@bernhardoj)

Note: I have only applied this on the timezone page, looks like we can't apply this on the currency selector in an easy way since we don't have a direct access the current selected currency.

### Results

https://user-images.githubusercontent.com/16493223/213521243-c17699cf-abfa-445f-a49d-645e04733153.mp4

**Comment by tienifr** [comment]
@s77rt looks like your proposal has quite similar approach to my proposals in [here](https://github.com/Expensify/App/issues/14314#issuecomment-1386372114) and [here](https://github.com/Expensify/App/issues/14314#issuecomment-1386404745) (the addition of ‘animated: true’ was also covered). The use of ‘keyForList’ was also covered [here](https://github.com/Expensify/App/issues/14314#issuecomment-1386365663).

Regarding the missing focus, I think that might be a conscious design decision since we already have the green check mark to indicate the selected option. You can check the Pronouns selection, it also follows the same behavior. 

If someone in the team agrees to fix this missing focus, I’d propose to fix in both Timezones and Pronouns as well by adding the ’indexOffset: 0’.

The correct behavior agreed in [here](https://github.com/Expensify/App/issues/14314#issuecomment-1397190657) is also missing in Pronouns and will need to be fixed similar to this.

**Comment by s77rt** [comment]
@tienifr The proposal may similar but not the same, you didn't provide a complete proposal. We do things a bit different e.g. the way we update the focused index.

**Comment by sketchydroide** [comment]
@tienifr your proposal does not scroll the list to current set timezone (at least I did not see it on the video)
> You can check the Pronouns selection

I think that is wrong as well, but I might need to ask design, I think it's good UX practice to show the current selected value

@s77rt your seems to work on the video

**Comment by tienifr** [comment]
@sketchydroide the scroll part is in the second part of my proposal [here](https://github.com/Expensify/App/issues/14314#issuecomment-1386444682). The video you see was [the part before that](https://github.com/Expensify/App/issues/14314#issuecomment-1386372114) so it didn't have the scroll. Sorry for the inconvenience

Let me paste the full diff for my existing proposals above for ease of review:
(I've also pushed it to [this branch](https://github.com/Expensify/App/compare/main...tienifr:App:fork/feat/14314) in case you want to pull and try it out)
```diff
diff --git a/src/components/OptionsSelector/BaseOptionsSelector.js b/src/components/OptionsSelector/BaseOptionsSelector.js
index c473c426d8..a5a4337d83 100755
--- a/src/components/OptionsSelector/BaseOptionsSelector.js
+++ b/src/components/OptionsSelector/BaseOptionsSelector.js
@@ -39,9 +39,18 @@ class BaseOptionsSelector extends Component {
         this.relatedTarget = null;
 
         const allOptions = this.flattenSections();
+
+        let focusedIndex = this.props.shouldTextInputAppearBelowOptions ? allOptions.length : 0;
+
+        const focusedOptionIndex = _.findIndex(allOptions, option => option.text === this.props.focusedValue);
+
+        if (focusedOptionIndex >= 0) {
+            focusedIndex = focusedOptionIndex;
+        }
+
         this.state = {
             allOptions,
-            focusedIndex: this.props.shouldTextInputAppearBelowOptions ? allOptions.length : 0,
+            focusedIndex,
         };
     }
 
@@ -93,6 +102,8 @@ class BaseOptionsSelector extends Component {
         } else {
             this.textInput.focus();
         }
+
+        this.scrollToIndex(this.state.focusedIndex, false);
     }
 
     componentDidUpdate(prevProps) {
@@ -175,8 +186,9 @@ class BaseOptionsSelector extends Component {
      * Scrolls to the focused index within the SectionList
      *
      * @param {Number} index
+     * @param {Boolean} animated
      */
-    scrollToIndex(index) {
+    scrollToIndex(index, animated = true) {
         const option = this.state.allOptions[index];
         if (!this.list || !option) {
             return;
@@ -195,7 +207,7 @@ class BaseOptionsSelector extends Component {
             }
         }
 
-        this.list.scrollToLocation({sectionIndex: adjustedSectionIndex, itemIndex});
+        this.list.scrollToLocation({sectionIndex: adjustedSectionIndex, itemIndex, animated});
     }
 
     /**
diff --git a/src/components/OptionsSelector/optionsSelectorPropTypes.js b/src/components/OptionsSelector/optionsSelectorPropTypes.js
index 306b89a26e..31150e7113 100644
--- a/src/components/OptionsSelector/optionsSelectorPropTypes.js
+++ b/src/components/OptionsSelector/optionsSelectorPropTypes.js
@@ -27,6 +27,9 @@ const propTypes = {
     /** Value in the search input field */
     value: PropTypes.string.isRequired,
 
+    /** Value of the option that we should focused on when first opening the options page */
+    focusedValue: PropTypes.string,
+
     /** Callback fired when text changes */
     onChangeText: PropTypes.func.isRequired,
 
diff --git a/src/pages/settings/Profile/TimezoneSelectPage.js b/src/pages/settings/Profile/TimezoneSelectPage.js
index 7a89ac8ce7..6ccffdcd85 100644
--- a/src/pages/settings/Profile/TimezoneSelectPage.js
+++ b/src/pages/settings/Profile/TimezoneSelectPage.js
@@ -61,6 +61,9 @@ class TimezoneSelectPage extends Component {
      */
     saveSelectedTimezone({text}) {
         PersonalDetails.updateSelectedTimezone(text);
+        this.setState({
+            timezoneInputText: text,
+        });
     }
 
     /**
@@ -88,8 +91,9 @@ class TimezoneSelectPage extends Component {
                     onChangeText={this.filterShownTimezones}
                     onSelectRow={this.saveSelectedTimezone}
                     optionHoveredStyle={styles.hoveredComponentBG}
-                    sections={[{data: this.state.timezoneOptions}]}
+                    sections={[{data: this.state.timezoneOptions, indexOffset: 0}]}
                     shouldHaveOptionSeparator
+                    focusedValue={this.state.timezoneInputText}
                 />
             </ScreenWrapper>
         );

```

> I'm using the option.text as the criteria to identify the selected option, but we might consider using other sorts of id (like keyForList) if we think it's more precise.

As mentioned we might consider using other sorts of id (like keyForList), but that's an implementation detail.

# After the fix

https://user-images.githubusercontent.com/113963320/213684102-2f6e7a14-6186-4706-bb99-7b1b275df500.mov

**Comment by sketchydroide** [comment]
@rushatgabhane what do you think?
Both @tienifr and @s77rt approaches are very similar

**Comment by shawnborton** [comment]
I like the idea of showing the selected item in the center of the view, aka scrolling to the right position when navigating back to the view with these selection lists (pronoun, currency, timezone, etc). However, I don't like how in [this proposal here](https://github.com/Expensify/App/issues/14314#issuecomment-1397379564) that the view seems to be moving a bit as it comes into view. 

cc @Beamanator too since I think you worked on implementing this?

**Comment by sketchydroide** [comment]
I noticed the small bump in both the proposals, I though it was the list loading, but that does not currently happen, must be something else, maybe the animated does not quite work as expected?

**Comment by s77rt** [comment]
I think it's related to FlatList/SectionList (a visualised list), the height of some items may have not been calculated yet. and scrolling into those items make the list calculate those heights correctly, I think it can be fixed by tweaking the parameters of rendering. (not tested)

**Comment by tienifr** [comment]
@sketchydroide  ok I found the issue

Our logic of `getItemLayout` is not 100% correct when the `shouldHaveOptionSeparator` is true. In that case there's a border top of 1px in all option (except the 1st one), but we're not accounting for that when calculating the offset.

The following will fix it:
(I've also pushed it to my branch of my proposal [here](https://github.com/Expensify/App/compare/main...tienifr:App:fork/feat/14314))
```diff
diff --git a/src/components/OptionsList/BaseOptionsList.js b/src/components/OptionsList/BaseOptionsList.js
index b7b6ebcd5a..95f54b5fca 100644
--- a/src/components/OptionsList/BaseOptionsList.js
+++ b/src/components/OptionsList/BaseOptionsList.js
@@ -121,6 +121,9 @@ class BaseOptionsList extends Component {
 
             // Add section items
             for (let i = 0; i < section.data.length; i++) {
+                if (i > 0 && this.props.shouldHaveOptionSeparator) {
+                    offset += variables.optionBorderTopWidth;
+                }
                 flatArray.push({length: optionHeight, offset});
                 offset += optionHeight;
             }
diff --git a/src/styles/variables.js b/src/styles/variables.js
index 9698c29ca8..be5c866d26 100644
--- a/src/styles/variables.js
+++ b/src/styles/variables.js
@@ -71,6 +71,7 @@ export default {
     popoverMenuShadow: '0px 4px 12px 0px rgba(0, 0, 0, 0.06)',
     optionRowHeight: 64,
     optionRowHeightCompact: 52,
+    optionBorderTopWidth: 1,
     optionsListSectionHeaderHeight: getValueUsingPixelRatio(54, 60),
     overlayOpacity: 0.6,
     lineHeightSmall: getValueUsingPixelRatio(14, 16),

```

# No more bump after the fix:
https://user-images.githubusercontent.com/113963320/213750369-f534a347-6324-4620-9925-e9500925321a.mov

cc @shawnborton

**Comment by shawnborton** [comment]
That does feel better.

**Comment by sketchydroide** [comment]
That makes sense @tienifr, thanks for finding that.

**Comment by tienifr** [comment]
@rushatgabhane there should be no impact since the `focusedValue` is a new field, other pages will not have it and will retain the same logic.

The `getItemLayout` change [here](https://github.com/Expensify/App/issues/14314#issuecomment-1398626471) will also not impact since it only makes the calculation correct. I think currently only for the scrolling issue above, the bug with the `getItemLayout` surfaces since it's scrolling through a lot of optionRow at once, so a `1px` difference for each row can cumulate and cause problems.

I tested with the Currency selection, Pronouns selection, New group creation, all are fine.

**Comment by s77rt** [comment]
### Proposal (Updated)
```diff
diff --git a/src/components/OptionRow.js b/src/components/OptionRow.js
index e25b0b807b..5d2486dccd 100644
--- a/src/components/OptionRow.js
+++ b/src/components/OptionRow.js
@@ -54,9 +54,6 @@ const propTypes = {
     /** Whether this option should be disabled */
     isDisabled: PropTypes.bool,
 
-    /** Whether to show a line separating options in list */
-    shouldHaveOptionSeparator: PropTypes.bool,
-
     style: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.object), PropTypes.object]),
 
     ...withLocalizePropTypes,
@@ -72,7 +69,6 @@ const defaultProps = {
     isDisabled: false,
     optionIsFocused: false,
     style: null,
-    shouldHaveOptionSeparator: false,
 };
 
 const OptionRow = (props) => {
@@ -134,7 +130,6 @@ const OptionRow = (props) => {
                             props.optionIsFocused ? styles.sidebarLinkActive : null,
                             hovered && !props.optionIsFocused ? props.hoverStyle : null,
                             props.isDisabled && styles.cursorDisabled,
-                            props.shouldHaveOptionSeparator && styles.borderTop,
                         ]}
                     >
                         <View style={sidebarInnerRowStyle}>
diff --git a/src/components/OptionsList/BaseOptionsList.js b/src/components/OptionsList/BaseOptionsList.js
index b7b6ebcd5a..2fe2543a7d 100644
--- a/src/components/OptionsList/BaseOptionsList.js
+++ b/src/components/OptionsList/BaseOptionsList.js
@@ -104,7 +104,10 @@ class BaseOptionsList extends Component {
      * @returns {Array<Object>}
      */
     buildFlatSectionArray() {
-        const optionHeight = variables.optionRowHeight;
+        let optionHeight = variables.optionRowHeight;
+        if (this.props.shouldHaveOptionSeparator) {
+            optionHeight += styles.borderTop.borderTopWidth;
+        }
         let offset = 0;
 
         // Start with just an empty list header
@@ -166,7 +169,6 @@ class BaseOptionsList extends Component {
                 showSelectedState={this.props.canSelectMultipleOptions}
                 boldStyle={this.props.boldStyle}
                 isDisabled={this.props.isDisabled || section.isDisabled}
-                shouldHaveOptionSeparator={index > 0 && this.props.shouldHaveOptionSeparator}
             />
         );
     }
@@ -231,6 +233,7 @@ class BaseOptionsList extends Component {
                     windowSize={5}
                     viewabilityConfig={{viewAreaCoveragePercentThreshold: 95}}
                     onViewableItemsChanged={this.onViewableItemsChanged}
+                    ItemSeparatorComponent={this.props.shouldHaveOptionSeparator && (() => <View style={styles.borderTop} />)}
                 />
             </View>
         );
diff --git a/src/components/OptionsSelector/BaseOptionsSelector.js b/src/components/OptionsSelector/BaseOptionsSelector.js
index c473c426d8..53298b8aa0 100755
--- a/src/components/OptionsSelector/BaseOptionsSelector.js
+++ b/src/components/OptionsSelector/BaseOptionsSelector.js
@@ -41,7 +41,7 @@ class BaseOptionsSelector extends Component {
         const allOptions = this.flattenSections();
         this.state = {
             allOptions,
-            focusedIndex: this.props.shouldTextInputAppearBelowOptions ? allOptions.length : 0,
+            focusedIndex: props.shouldTextInputAppearBelowOptions ? allOptions.length : 0,
         };
     }
 
@@ -84,6 +84,10 @@ class BaseOptionsSelector extends Component {
             true,
         );
 
+        if (this.props.initiallyFocusedItemKey) {
+            this.updateFocusedIndex(_.findIndex(this.state.allOptions, option => option.keyForList === this.props.initiallyFocusedItemKey), false);
+        }
+
         if (!this.props.autoFocus) {
             return;
         }
@@ -166,17 +170,19 @@ class BaseOptionsSelector extends Component {
 
     /**
      * @param {Number} index
+     * @param {Boolean} animated
      */
-    updateFocusedIndex(index) {
-        this.setState({focusedIndex: index}, () => this.scrollToIndex(index));
+    updateFocusedIndex(index, animated = true) {
+        this.setState({focusedIndex: index}, () => this.scrollToIndex(index, animated));
     }
 
     /**
      * Scrolls to the focused index within the SectionList
      *
      * @param {Number} index
+     * @param {Boolean} animated
      */
-    scrollToIndex(index) {
+    scrollToIndex(index, animated = true) {
         const option = this.state.allOptions[index];
         if (!this.list || !option) {
             return;
@@ -195,7 +201,7 @@ class BaseOptionsSelector extends Component {
             }
         }
 
-        this.list.scrollToLocation({sectionIndex: adjustedSectionIndex, itemIndex});
+        this.list.scrollToLocation({sectionIndex: adjustedSectionIndex, itemIndex, animated});
     }
 
     /**
diff --git a/src/components/OptionsSelector/optionsSelectorPropTypes.js b/src/components/OptionsSelector/optionsSelectorPropTypes.js
index 306b89a26e..8940b63ab5 100644
--- a/src/components/OptionsSelector/optionsSelectorPropTypes.js
+++ b/src/components/OptionsSelector/optionsSelectorPropTypes.js
@@ -89,6 +89,9 @@ const propTypes = {
 
     /** Whether to show a line separating options in list */
     shouldHaveOptionSeparator: PropTypes.bool,
+
+    /** Initially focused item key */
+    initiallyFocusedItemKey: PropTypes.string,
 };
 
 const defaultProps = {
@@ -113,6 +116,7 @@ const defaultProps = {
     disableArrowKeysActions: false,
     isDisabled: false,
     shouldHaveOptionSeparator: false,
+    initiallyFocusedItemKey: '',
 };
 
 export {propTypes, defaultProps};
diff --git a/src/pages/settings/Profile/TimezoneSelectPage.js b/src/pages/settings/Profile/TimezoneSelectPage.js
index 7a89ac8ce7..0297113ecd 100644
--- a/src/pages/settings/Profile/TimezoneSelectPage.js
+++ b/src/pages/settings/Profile/TimezoneSelectPage.js
@@ -88,7 +88,8 @@ class TimezoneSelectPage extends Component {
                     onChangeText={this.filterShownTimezones}
                     onSelectRow={this.saveSelectedTimezone}
                     optionHoveredStyle={styles.hoveredComponentBG}
-                    sections={[{data: this.state.timezoneOptions}]}
+                    sections={[{data: this.state.timezoneOptions, indexOffset: 0}]}
+                    initiallyFocusedItemKey={this.currentSelectedTimezone}
                     shouldHaveOptionSeparator
                 />
             </ScreenWrapper>
```

### Update
- Renamed `focusedItemKey` to `initiallyFocusedItemKey`
- Removed the extra prop and style that is passed to `<OptionRow />` We should instead use FlatList/SectionList `ItemSeparatorComponent`
- Corrected the height calculation. @tienifr proposal above only updates the offset and not the `optionHeight` (that is pushed to the array)

___

Still waiting for a confirmation to apply the focus fix on other lists.

**Comment by tienifr** [comment]
@s77rt it was intended that the offset is updated rather than the height. Technically the border is not considered part of the option height.

**Comment by s77rt** [comment]
Actually it's a part of it. Borders are parts of the elements. There is an alternative which outline but it does not support edges. So we have to stick with borders and I think it's better to use the virtualized list item separator component after all.

**Comment by tienifr** [comment]
@s77rt in CSS you’d have to specify a property box-sizing to indicate if your width/height includes the border (it excludes border by default). So the border is not part of the item height for me.
Would love to hear others’ thoughts too!

**Comment by s77rt** [comment]
I think we use border-box already. You can check the height of items yourself, you will find it's 65px not 64px.
Anyway, we should not apply that style to the raw `<OptionRow />` in the first place we should instead use the item separator.

**Comment by tienifr** [comment]
@s77rt just a small note `you will find it's 65px not 64px` means we’re not using `box-sizing: border-box`, if we use it we’ll see the height is 64 even with the border, that’s how the property works 👍

**Comment by s77rt** [comment]
@tienifr Actually it's the other way around. Using border-box means that the height includes the border: `height = content height + padding + border`. Attached a video to confirm that and to confirm that we are using `box-sizing: border-box`.

https://user-images.githubusercontent.com/16493223/213861927-40affc1e-56b8-4e61-9c4d-0db5d617bf96.mp4

___

> Also, if you consider the border to be the optionHeight then the first row height will be 64 (your proposal says 65) since it has no border and the rest 65 -> inconsistent.

Actually that's the case with your proposal not mine. You can see that the first element does not have the border style thus it's 64px while the rest 65px.

With my proposal all elements have the same height of 65px and still the border (separator) is only applied when necessary That makes it's easier for the virtualized list to scroll to any element since they all have same height now.

**Comment by s77rt** [comment]
Here with my proposal using `ItemSeparatorComponent` all items having a height of 65px.

https://user-images.githubusercontent.com/16493223/213862379-161647b6-adac-49e8-9453-8adf269a86c8.mp4

**Comment by tienifr** [comment]
@s77rt Your video in [here](https://github.com/Expensify/App/issues/14314#issuecomment-1399221326) is not a good example of how `border-box` works since:
- The element you show doesn't have the `height` property
- The element you show doesn't have any border

Let me give you a better example in the below video:
- In this video the component has all height, border and box-sizing property
- When we use `border-box`, the height you see is 64, since the `border` is already included in the `height`
- When we use `content-box`, the height you see is 64.99 (~65), since what you see in the browser is composed of the element "content" height that you defined and the border

That aligns with what I said in [here](https://github.com/Expensify/App/issues/14314#issuecomment-1399187365). In case of a proper example, the height you see with `border-box` will always be <= the height you see with the default `content-box`.

https://user-images.githubusercontent.com/113963320/213869474-08410d4f-67eb-40a7-bc20-2bb82ff22b36.mov

This doesn't really relate much to this issue, but hope it helps.

**Comment by tienifr** [comment]
> I think it's better to use the virtualized list item separator component after all.

I think using it is ok, if we don't mind having a bunch of additional DOM Nodes (for the `ItemSeparatorComponent`, roughly as many as the number of elements) lying around. The current implementation works just fine with the change in the offset calculation.

**Comment by s77rt** [comment]
@tienifr Sorry but I think I have made my case regarding `box-sizing` clear already thus I don't have much to add just to mention that you are inspecting the wrong element, you should inspect the list items not the children of the item. The height are calculated using the list items themselves.

___

The `length` (`optionHeight`) and `offset` are pretty much linked to each other, the way you are doing it does not make much sense. If in doubt you can check here https://github.com/Expensify/App/issues/12772 where we dealt with height calculation issue.
___

Using `ItemSeparatorComponent` is the react native way.

**Comment by s77rt** [comment]
@rushatgabhane The mdn docs is actually what I'm saying and that translates to: `height: content height + border + padding`. In a way you can view this from two perspective, part of @tienifr is actually right but since he overwrote the css and added `height` himself which does not exist in our App things look bit different. You can verify, inspect the same element, it's height is 65px, and @tienifr overwrote it to look like it's 64px just to prove his point.

Here I will explain it in a straightforward way (inspecting the same element @tienifr did inspect), the item is using `border-box` so the height that we define should be calculated using this formula: `height: content height + border + padding`

In the attached video you can see that:
content height: 64px
padding height: 0
border height: 1px
==> height: 65px


https://user-images.githubusercontent.com/16493223/214020508-03a64686-8661-4b8a-8942-e670e0e216bd.mp4

**Comment by s77rt** [comment]
In fact, you don't even need to get into `box-border` calculations, you can just take the computed style and use it in `getItemLayout` and the computed height is 64px for the first item and 65px for the rest as seen here https://github.com/Expensify/App/issues/14314#issuecomment-1399221326 where @tienifr proposal accounts that as 64px for every item.

In my proposal things are a bit different, where every item is 65px (as you can see here https://github.com/Expensify/App/issues/14314#issuecomment-1399222497) and I account for those correctly as 65px

**Comment by tienifr** [comment]
@s77rt in order to validate if your example is correct, you can try adding a ‘box-sizing: content-box’ to that same element that you’re showing in the video. You’ll probably see that the height is still 65 with both values of box-sizing, which proves that the 65 comes from something else, not the border-box property (aka the example isn’t a good one that shows what border-box does)

**Comment by s77rt** [comment]
@tienifr `box-sizing: border-box` is applied in the entire app. You can't just change one element and expect a huge change. I think it would be more clear to focus on what we have (`box-sizing: border-box`) and forget about `content-box` as it's irrelevant.

I think it's crystal clear that the element height is 65px (first item) and 64px (rest of items). If [that video](https://github.com/Expensify/App/issues/14314#issuecomment-1399221326) is not convincing, I don't know what will.

**Comment by sketchydroide** [comment]
Thanks for the clarification, but I think @tienifr proposal makes sense as well, and agree with @parasharrajat, so I assigned @tienifr

**Comment by s77rt** [comment]
@sketchydroide The height is not calculated correctly

**Comment by tienifr** [comment]
Thanks @sketchydroide @rushatgabhane , I've applied to the job. The PR will be ready within 1 day.

**Comment by tienifr** [comment]
@rushatgabhane @sketchydroide the PR is ready for review https://github.com/Expensify/App/pull/14466
Thanks

**Comment by Pujan92** [comment]
IMO partially I agree to both @s77rt & @tienifr 

-  box-sizing property include the padding and border in an element's total height, but here I am not seeing it is relevant as border is applied to parent element(TouchableOpacity) and not with the child element(View) where we applies height to `64px`.
- Use of`ItemSeparatorComponent` is a good way to use separator as it won't be included in the option element(currenlty a border is a part of the option row)

>        let optionHeight = variables.optionRowHeight;
>        if (this.props.shouldHaveOptionSeparator) {
>            optionHeight += styles.borderTop.borderTopWidth;
>        }

@s77rt I think we don't require to add the border height to the optionHeight when we are using the `ItemSeparatorComponent`, only for the offset calculation it needs to consider.

**Comment by s77rt** [comment]
@Pujan92 We should account for the border height in the `optionHeight` whether we are using `ItemSeparatorComponent` or not. The use of `ItemSeparatorComponent` is only an optimisation here.

**Comment by Pujan92** [comment]
> @Pujan92 Actually you may be right, using `ItemSeparatorComponent` only requires us to update the `offset`. But the current code it's not being used and thus we should account for the border on the height calculatation.

Yes, To me if we won't consider `ItemSeparatorComponent` then the selected proposal seems fine for the fix.

**Comment by s77rt** [comment]
@Pujan92 Am I missing something? The proposal does not update the `optionHeight` accordingly

**Comment by Pujan92** [comment]
> @Pujan92 Am I missing something? The proposal does not update the `optionHeight` accordingly

Oh, Yes. I think in that case `optionHeight` also needs to add that border which is added to the offset.

**Comment by kadiealexander** [comment]
@rushatgabhane have you reviewed all proposals here? Do you think it's time to increase the price?

**Comment by situchan** [comment]
Is this still open to proposals? I think @tienifr was already assigned [here](https://github.com/Expensify/App/issues/14314#issuecomment-1400205298) and thus `Help Wanted` label removed.

**Comment by sketchydroide** [comment]
@kadiealexander what @situchan said, we had already assigned @tienifr, no need to update price

**Comment by kadiealexander** [comment]
Ignore me 🤦 sorry team! I missed the comment and was confused by the continuing discussion.

**Comment by tienifr** [comment]
@rushatgabhane As I can see in https://github.com/Expensify/App/issues/14781#issuecomment-1416526684 and I also tested my PR,I think my PR doesn't cause the regression. What do you mean 'an updated proposal'?. Or I can simply create the same PR as before. Thanks

**Comment by thienlnam** [comment]
@tienifr Please just try to re-create the PR first and see if it adds a regression of this https://github.com/Expensify/App/issues/14781

There's also a chance that it might not have and there was just [something weird](https://github.com/Expensify/App/issues/14781#issuecomment-1419600245) with staging at the time

**Comment by tienifr** [comment]
@thienlnam @rushatgabhane 

# Regression Test Proposal

**Bug**: Invite new members - Users is highlighted after selecting

**Proposed Test Steps**:

- Open the app
- Open Workspace and go to the invite member page
- Select some members
- Verify that selected members aren't highlighted

Do we :+1: or :-1: 

Video (Regression Test):
https://user-images.githubusercontent.com/113963320/217418140-3cb8ccba-cdf6-4c41-b674-cd51628590af.mov

**Comment by tienifr** [comment]
@thienlnam @rushatgabhane I added regression tests for both issues https://github.com/Expensify/App/issues/14314 and https://github.com/Expensify/App/issues/14781. Please help to check again

**Comment by tienifr** [comment]
@thienlnam @rushatgabhane 

# Regression Test Proposal

**Bug**: Timezone changes to some 1st option in the list while pressing the enter key

**Proposed Test Steps**:

1. Go to Profile and click on timezone
2. Change timezone to some other regions and go back
3. Again click on timezone and now hit the enter key
4. Notice that timezone changes to some other region on pressing the enter key

Do we :+1: or :-1: 

Video (Regression Test):

https://user-images.githubusercontent.com/113963320/218412461-8bdb09e8-9d6b-46a4-b470-1e9e5fc6c2c3.mp4

**Comment by MelvinBot** [comment]
@sketchydroide, @rushatgabhane, @kadiealexander, @tienifr Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by Beamanator** [comment]
@thienlnam do you want to take over this issue?

**Comment by tienifr** [comment]
@Beamanator The PR is created again and it is being reviewed by @rushatgabhane

**Comment by Beamanator** [comment]
@tienifr thanks, looks like you're talking about https://github.com/Expensify/App/pull/14938?

I think @thienlnam or I can take over from the expensify side

**Comment by tienifr** [comment]
> thanks, looks like you're talking about https://github.com/Expensify/App/pull/14938?

@Beamanator yes

I think we can just speed up reviewing [that PR](https://github.com/Expensify/App/pull/14938), it has the same changes as [the one approved and merged previously](https://github.com/Expensify/App/pull/14466). It didn't cause the regression but was reverted by mistake.

The regression tests for the related issues were already added as well.

cc @rushatgabhane @thienlnam

**Comment by thienlnam** [comment]
@Beamanator Yeah, I can take over this issue

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.76-7 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/14938


If no regressions arise, payment will be issued on 2023-03-06. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@rushatgabhane / @thienlnam] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@rushatgabhane / @thienlnam] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@rushatgabhane / @thienlnam] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@tienifr] Propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@kadiealexander] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by tienifr** [comment]
I created regression test before

> # Regression Test Proposal
> 
> **Bug**: Timezone changes to some 1st option in the list while pressing the enter key
> 
> **Proposed Test Steps**:
> 
> 1. Go to Profile and click on timezone
> 2. Change timezone to some other regions and go back
> 3. Again click on timezone and now hit the enter key
> 4. Notice that timezone changes to some other region on pressing the enter key
> 
> Do we :+1: or :-1: 
> 
> Video (Regression Test):
> 
> https://user-images.githubusercontent.com/113963320/218412461-8bdb09e8-9d6b-46a4-b470-1e9e5fc6c2c3.mp4
>

**Comment by priya-zha** [comment]
@kadiealexander Upwork doesn't allow me to apply for this job. I get an error.
![Screenshot from 2023-03-09 09-02-20](https://user-images.githubusercontent.com/34065075/223907960-d06a66a5-21ad-4caf-aeb3-7e085cfaa3e9.png)

**Comment by MelvinBot** [comment]
📣 @priya-zha! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by kadiealexander** [comment]
@priya-zha could you please complete the steps above first?

**Comment by priya-zha** [comment]
Contributor details
Your Expensify account email: pj35134@gmail.com
Upwork Profile Link:https://www.upwork.com/freelancers/~011e32317dbbd489e0

**Comment by kadiealexander** [comment]
Thanks @priya-zha! I have sent you a contract in Upwork, please accept and we'll issue payment.

**Comment by kadiealexander** [comment]
Awesome, all paid! Payment issued to @tienifr too. Just waiting for @rushatgabhane to accept. 

@thienlnam are you happy with the regression test proposed [here](https://github.com/Expensify/App/issues/14314#issuecomment-1447679112)?

**Comment by thienlnam** [comment]
Sorry for the delay on this - could we please try to be more specific in the regression tests so it's easier for our QA team to follow? @tienifr 

Things like updating this 
> Change timezone to some other regions and go back

to: 
Make sure the 'Automatically determine your location' toggle is turned off so you can update your timezone manually
Update the timezone to a random region and then go back to the profile page

Also, we want to test the success case so :

> Notice that timezone changes to some other region on pressing the enter key

We don't want this to happen right? So

> Verify that the timezone does not change to another region after pressing the enter key

**Comment by tienifr** [comment]
# Regression Test Proposal

**Bug**: Timezone changes to some 1st option in the list while pressing the enter key

**Proposed Test Steps**:

1. Go to Profile and click on timezone
2. Ensure that you have disabled the 'Automatically determine your location' option so that you can manually adjust your timezone. 
3. Change the timezone to a different location and then return to the profile page.
4. Click again on timezone and now hit the enter key
5. Verify that the timezone does not change to another region after pressing the enter key
Do we :+1: or :-1: 

Video (Regression Test):

https://user-images.githubusercontent.com/113963320/218412461-8bdb09e8-9d6b-46a4-b470-1e9e5fc6c2c3.mp4

**Comment by thienlnam** [comment]
Awesome, looks great now - thanks!

**Comment by kadiealexander** [comment]
https://github.com/Expensify/App/issues/15940 - GH to get regression test added to TestFlight. @thienlnam is there anything else pending before we close this out?
